pkg/adapter/awscognitouserpoolsource/adapter_test.go

```diff
@@ -0,0 +1,150 @@
+	cognitoidentityprovideriface.CognitoIdentityProviderAPI
+	listUsersOutput cognitoidentityprovider.ListUsersOutput
+}
+
+func (m mockedCognitoUserPoolClient) ListUsers(in *cognitoidentityprovider.ListUsersInput) (*cognitoidentityprovider.ListUsersOutput, error) {
+	return &m.listUsersOutput, nil
+}
+
+func TestListUsers(t *testing.T) {
+	user := cognitoidentityprovider.UserType{
+		UserLastModifiedDate: func() *time.Time {
+			now := time.Now()
+			return &now
+		}(),
+	}
+
+	a := &adapter{
+		userPoolID: "userpool/fooPool",
+		logger:     loggingtesting.TestLogger(t),
+		ceClient:   adaptertest.NewTestClient(),
+		cgnIdentityClient: mockedCognitoUserPoolClient{
+			listUsersOutput: cognitoidentityprovider.ListUsersOutput{
+				Users: []*cognitoidentityprovider.UserType{
+					&user,
+				},
+			},
+		},
+	}
+
+	users, err := a.listUsers()
+	assert.NoError(t, err)
+	assert.Equal(t, []*cognitoidentityprovider.UserType{&user}, users)
+}
+
+func TestFilterByTimestamp(t *testing.T) {
+	now := time.Now()
+
+	// existing user
+	user1 := cognitoidentityprovider.UserType{
+		UserLastModifiedDate: &now,
+	}
+
+	// we must get 0 new users
+	users, timestamp := filterByTimestamp([]*cognitoidentityprovider.UserType{
+		&user1,
+	}, now)
+	assert.Len(t, users, 0)
+
+	// new user
+	user2 := cognitoidentityprovider.UserType{
+		UserLastModifiedDate: func() *time.Time {
+			now := time.Now()
+			return &now
+		}(),
+	}
+
+	// we must get user2 only
+	users, timestamp = filterByTimestamp([]*cognitoidentityprovider.UserType{
+		&user1,
+		&user2,
+	}, timestamp)
+	assert.Equal(t, []*cognitoidentityprovider.UserType{&user2}, users)
+
+	// repeat filtering, we must ignore user2
+	users, _ = filterByTimestamp([]*cognitoidentityprovider.UserType{
+		&user1,
+		&user2,
+	}, timestamp)
+	assert.Len(t, users, 0)
+}
+
+func TestSendCognitoEvent(t *testing.T) {
+	ceClient := adaptertest.NewTestClient()
+
+	user := cognitoidentityprovider.UserType{
+		UserLastModifiedDate: func() *time.Time {
+			now := time.Now().UTC()
+			return &now
+		}(),
+	}
+
+	a := &adapter{
+		userPoolID:        "userpool/fooPool",
+		logger:            loggingtesting.TestLogger(t),
+		ceClient:          ceClient,
+		cgnIdentityClient: mockedCognitoUserPoolClient{},
+	}
+
+	err := a.sendCognitoEvent(&user)
+	assert.NoError(t, err)
+
+	events := ceClient.Sent()
+	assert.Len(t, events, 1)
+
+	var gotUser cognitoidentityprovider.UserType
+	err = events[0].DataAs(&gotUser)
+	assert.NoError(t, err)
+	// make sure that we get what we sent
+	assert.Equal(t, user, gotUser)
+}
+
+func TestStart(t *testing.T) {
+	a := &adapter{
+		userPoolID:        "userpool/fooPool",
+		logger:            loggingtesting.TestLogger(t),
+		ceClient:          adaptertest.NewTestClient(),
+		cgnIdentityClient: mockedCognitoUserPoolClient{},
+	}
+
+	stopCh := make(chan struct{})
+	go func() {
+		close(stopCh)
+	}()
+
+	// test graceful shutdown
+	err := a.Start(stopCh)
+	assert.NoError(t, err)
```

Comment:
Your `Start()` function may never react to the signal, and then your test will run until it times out (10 min). Here's a suggestion to ensure we time out quickly (after 2s):

```go
	// create a context that will be done in 2s
	testCtx, testCancel := context.WithTimeout(context.Background(), 2*time.Second)
	defer testCancel()  // not really necessary here, but let's follow good practices

	// channel to receive stop signals
	stopCh := make(chan struct{})

	// channel to receive the value returned by Start()
	errCh := make(chan error)

	// run adapter handler in the background
	go func() {
		errCh <- a.Start(stopCh)
	}()

	// simulate sending a stop signal to the adapter
	close(stopCh)

	// if the timeout is done before the handler returned, the test is always failed
	select {
	case <-testCtx.Done():
		t.Errorf("Test timed out after %v", testTimeout)
	case err := <-errCh:
		assert.NoErrorf(t, err, "Receiver returned an error")
	}
```